pkg/vault/contracts/Vault.sol

```diff
@@ -1185,43 +1187,54 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
             _bufferTokenBalances[wrappedToken] = bufferBalances;
         } else {
             // The buffer does not have enough liquidity to facilitate the wrap without making an external call.
-            // We wrap the user's tokens via an external call and additionally rebalance the buffer if it has a
-            // surplus of underlying tokens.
+            // We wrap the user's tokens via an external call and additionally rebalance the buffer if it has an
+            // imbalance of underlying tokens.
 
-            // Gets the amount of underlying to wrap in order to rebalance the buffer.
-            uint256 bufferUnderlyingSurplus = bufferBalances.getBufferUnderlyingSurplus(wrappedToken);
-            uint256 bufferWrappedSurplus;
+            // Gets the imbalance of underlying in the buffer. If negative, the buffer has more wrapped than underlying.
+            int256 bufferUnderlyingImbalance = bufferBalances.getBufferUnderlyingImbalance(wrappedToken);
+            int256 bufferWrappedImbalance = bufferBalances.getBufferWrappedImbalance(wrappedToken);
 
-            // Expected amount of underlying deposited into the wrapper protocol.
-            uint256 vaultUnderlyingDeltaHint;
-            // Expected amount of wrapped minted by the wrapper protocol.
-            uint256 vaultWrappedDeltaHint;
+            // Expected amount of underlying deposited into the wrapper protocol if buffer is balanced.
+            uint256 vaultUnderlyingDeltaHint = amountInUnderlying;
+            // Expected amount of wrapped minted by the wrapper protocol if buffer is balanced.
+            uint256 vaultWrappedDeltaHint = amountOutWrapped;
```

Comment:
Actually, it doesn't. For ExactOut, the if had a typo, it should be checking the imbalance and not the delta hint. So, for ExactOut, if there's no imbalance in the buffer, these values will be used to wrap.

Reply:
hm, got it.

I have mixed feelings about that last `if (bufferUnderlyingImbalance != 0)`. It does optimize things for buffers that have very little liquidity. Perhaps in some L2s buffers will be initialized just for the system to work, but without any meaningful liquidity. In that case, there's no need to rebalance it ever, as any trade will always just go to the lending protocol, and the buffer will always stay balanced. More concretely, it saves one `preview` function which can be interesting (the sum doesn't move the needle). I understand this case is tested in `testMintBufferBalancedNotEnoughLiquidity` under `BufferVaultPrimitive`.

On the other hand, it's pretty deep already (if not enough liquidity, if exact out, if rebalance is needed), and it introduces an asymmetry with exact in (which to be fair happens rather often with exact outs). And also the fact that we need to initialize `vaultUnderlyingDeltaHint` and `vaultWrappedDeltaHint` just in case we don't enter the most inner `if`, but at the same time we overwrite those two almost every time (i.e. unless it's perfectly balanced) is a bit tricky.

Without the last `if`, we have:
- if no liquidity --> creates delta hints, no assignment needed
  - exact in --> always sets delta hints 
  - exact out --> always sets delta hints 

and the worst part is that extra `preview` call. The result should be the same if we do the math right? i.e. for `bufferUnderlyingImbalance == 0` we get `vaultUnderlyingDeltaHint == amountInUnderlying` (same as now) and `vaultWrappedDeltaHint == previewDeposit(amountInUnderlying)` --> is this the same as `amountOutWrapped` or does that have a rounding error?

I'd love to hear somebody else's thoughts about this. This smells like an optimization that bites you back; I'm not sure it's really worth it.

Reply:
> `previewMint(previewDeposit(x)) == x`

This is not the case. That's actually a small source of rounding errors (which actually will happen almost always, since whenever we go to the lending protocol for liquidity chances are that the buffer is not going to be balanced).